In [1]:
# initial setup
%run "../../../common/0_notebooks_base_setup.py"


/media/paulati/Nuevo vol/paula/dh/2021/dsad_2021_paula/common
default checking
Running command `conda list`... ok
jupyterlab=2.2.6 already installed
pandas=1.1.5 already installed
bokeh=2.2.3 already installed
seaborn=0.11.0 already installed
matplotlib=3.3.2 already installed
ipywidgets=7.5.1 already installed
pytest=6.2.1 already installed
chardet=4.0.0 already installed
psutil=5.7.2 already installed
scipy=1.5.2 already installed
statsmodels=0.12.1 already installed
scikit-learn=0.23.2 already installed
xlrd=2.0.1 already installed
Running command `conda install --yes nltk=3.5.0`... ok
Solving environment: ...working... done

# All requested packages already installed.


unidecode=1.1.1 already installed
pydotplus=2.0.2 already installed
pandas-datareader=0.9.0 already installed
flask=1.1.2 already installed


In [2]:
import pandas as pd
import numpy as np
import seaborn as sns

<link rel="stylesheet" href="../../../common/dhds.css">
<div class="Table">
    <div class="Row">
        <div class="Cell grey left"> <img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_portada.png" align="center" width="90%"/></div>
        <div class="Cell right">
            <div class="div-logo"><img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/common/logo_DH.png" align="center" width=70% /></div>
            <div class="div-curso">DATA SCIENCE</div>
            <div class="div-modulo">MÓDULO 3</div>
            <div class="div-contenido">Regularización</div>
        </div>
    </div>
</div>

### Agenda

---

- Entender la regularización como técnica para evitar el sobreajuste

- Aplicar regularización usando scikit-learn con  validación cruzada para el ajuste de hiperparámetros.


<div class="div-dhds-fondo-1"> Sesgo - Varianza
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_separador.png" align="center" />

</div>

### Sesgo-varianza

---

Dado el modelo:

<p style="font-size:20px;">
$$Y = f(x) + \epsilon$$
</p>

Donde epsilon es un término aleatorio con distribución:

<p style="font-size:20px;">
$$\epsilon \sim N(0, \sigma_{\epsilon})$$
</p>
    
Podemos obtener una estimación de f para hacer predicciones sobre Y:

<p style="font-size:20px;">
$$\hat{Y} = \hat{f}(X)$$
</p>


La esperanza del error de predicción al cuadrado será la siguiente:

<p style="font-size:20px;">
$$Err(x) = E[(Y - \hat{f}(x))^2]$$
</p>

Podemos descomponer la esperanza del error al cuadrado de la siguiente manera: 

<p style="font-size:20px;">
$$Err(x) = (E[ \hat{f}(x)]-f(x))^2 + E[(\hat{f}(x) - E[\hat{f}(x)])^2] + \sigma_{\epsilon}^2$$
</p>

<p style="font-size:20px;">
$$Err(x) = Bias^2 + Variance + Irreducible \space error$$
</p>

Si el modelo es **demasiado simple** (tiene pocos grados de libertad), entonces no importa cuán grande sea la muestra tenemos **sesgo o error sistemático**:

$$E[(\hat{f}(x)] \ne E[(f(x)]$$

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_sesgo.png" align="center"/>


Si el modelo es **demasiado complejo** (ie. tiene demasiados grados de libertad), entonces el estimador puede ajustar regularidades espurias de la muestra generando **sobreajuste**

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_overfitting.png" align="center"/>

Por lo tanto, el modelo no debe ser ni muy simple ni muy complejo


<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_sesgo_varianza.png" align="center"/>

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_dilema_sesgo_varianza.png" align="center"/>

<div class="div-dhds-fondo-1"> Regularización
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_separador.png" align="center" />

</div>

### Regularización: una primera intuición

---

Función de pérdida de una regresión lineal:

<p style="font-size:20px;">
$$CF = \sum_{i=1}^N(\hat{y}_i - y_i)^2$$
</p>

Las técnicas de regularización agregan una **"penalidad"** a esta función de costo:


    
<p style="font-size:20px;">
$$CF = \sum_{i=1}^N(\hat{y}_i - y_i)^2 + \alpha . \vec{\theta}$$
</p>

$\vec{\theta}$ (theta) es el vector que corresponde a los parámetros del modelo (en una regresión lineal, los betas) y alpha es un parámetro que “regula” la fuerza de la penalización: cuanto más grande es, mayor es la penalización.

Vamos a ver a continuación dos técnicas de regularización

**Regresión Ridge**

**Regresión Lasso**

Estas técnicas proponen cambiar ligeramente el problema de optimización de mínimos cuadrados, para intentar **"achicar" (shrink) el valor absoluto de los estimadores de los betas $\beta_i$**. 


**¿Por qué esto mejoraría la estimación?** 

Vamos a ver de qué forma este método **introduce un sesgo pero reduce la varianza**.


### Normas L1 y L2

---

### Norma L0

La cantidad de elementos distintos de cero en el vector.


#### Norma L1

La suma de los valores absolutos de los elementos del vector

$\lVert \beta \rVert_1 = \sum_i \vert x_i \vert$


#### Norma L2

La raíz cuadrada de la suma de cuadrados de los elementos del vector

$\lVert \beta \rVert_2 = \sqrt{\sum_i  x_i^2 }$




#### Normalización L1

La idea de esta normalización es que la suma del valor absoluto sea unitaria:

$\sum_{j=1}^p \lVert x_{ij} \rVert = 1 \space \space \forall i = 1 .. n$

#### Normalización L2

La idea de esta normalización es que la suma del valor absoluto al cuadrado sea unitaria

$\sum_{j=1}^p \lVert x_{ij} \rVert^2 = 1 \space \space \forall i = 1 .. n$


### Regresión Ridge

---

Recordemos la función que se minimiza en la estimación de mínimos cuadrados:

<p style="font-size:20px;">
$$RSS = \sum_{i=1}^n (y_i - \beta_0 - \sum_{j=1}^p \beta_j x_{ij})^2  $$
</p>

La función que se minimiza en Regresión Ridge es:

<p style="font-size:20px;">
$$\sum_{i=1}^n (y_i - \beta_0 - \sum_{j=1}^p \beta_j x_{ij})^2  + \lambda \sum_{j=1}^p \beta_j^2 = RSS + \lambda \sum_{j=1}^p \beta_j^2$$
</p>

La diferencia es que agregamos un término nuevo. En este término, un **hiperparámetro $\lambda$** (lambda) penaliza el valor de los coeficientes al cuadrado. 
Entonces, tenemos que minimizar el cuadrado de los errores, intentando que ningún $\beta_j^2$ sea demasiado grande.



<p style="font-size:20px;">
$$\sum_{i=1}^n (y_i - \beta_0 - \sum_{j=1}^p \beta_j x_{ij})^2  + \lambda \sum_{j=1}^p \beta_j^2 = RSS + \lambda \sum_{j=1}^p \beta_j^2$$
</p>


- Al igual que MCO, buscamos achicar la función de pérdida.

- Existe un **término de penalización**, que es menor cuando los betas se acercan a cero, por lo tanto tiene el efecto de achicar los mismos hacia cero (tanto si son negativos como positivos)

- El **hiperparámetro $\lambda$** maneja la ponderación de cada término.

    ¿Cuál es el mejor valor para lambda? ¿Cómo elegíamos el valor óptimo de un hiperparámetro?

    Como siempre, lo hacemos mediante **CROSS VALIDATION**


### Ejemplo Ridge. Dataset Credit.

<img src = "https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_credit_ridge.png" align="center" />


La línea negra sólida representa la estimación de la regresión de ridge para el coeficiente de `income` en función de $\lambda$. 


En el extremo izquierdo del gráfico, $\lambda$ es esencialmente cero, por lo que las estimaciones de coeficientes de ridge correspondientes son las mismas que las estimaciones de mínimos cuadrados habituales. 


A medida que aumenta $\lambda$, las estimaciones del coeficiente de ridge se reducen a cero. Cuando $\lambda$ es extremadamente grande, entonces todas las estimaciones del coeficiente de ridge son básicamente cero; esto corresponde al modelo nulo que no contiene predictores.

En el panel de la derecha, podemos pensar el eje x como la cantidad en que se ha reducido el tamaño de las estimaciones del coeficiente de regresión de ridge.


### Regresión Ridge: Sensibilidad a la escala

---

- En Regresión Ridge tanto la estimación de los coeficientes como la predicción son **sensibles a la escala**.

- Si una variable se encuentra en una escala que le da un valor absoluto mayor, esto **va a afectar el cálculo de la suma de cuadrados del vector de coeficientes**.

- Por esta razón **es importante estandarizar (dividir por el desvío estándar)** todos los regresores antes de ejecutar una regresión Ridge. Así ya no están en unidades físicas sino en unidades de su propio desvío estándar.

<p style="font-size:20px;">
$$\tilde{x_{ij}} = \frac{x_{ij}}{\sqrt{\frac{1}{n}\sum_{i=1}^n(x_{ij} - \bar{x_j})^2}}$$
</p>


### Tradeoff Bias-Variance

---

Aquí se pueden ver n=50 simulaciones, p=45 predictores, todos con coeficientes no nulos.

El gráfico expresa el sesgo cuadrado (negro), varianza (verde) y el MSE del test (violeta), para una regresión ridge en los datos simulados, como una función de $\lambda$  y de 
<p style="font-size:20px;">
$\frac{ \lVert{\hat{\beta}_{\lambda}^R}\rVert_2}{\lVert{\hat{\beta}}\rVert_2}$
</p>

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_tradeoff_bias_variance.png" align="center" />


### Regresión Lasso

---

* La regresión Ridge tiene una clara desventaja: incluye todos los predictores p en el modelo final, a diferencia de aquellos modelos que eligen un subconjunto de predictores.

* La regresión Lasso es un alternativa a Ridge, que corrige esta desventaja. Los  coeficientes $\hat{\beta_{\lambda}^L}$ minimizan el número de predictores


* Lasso utiliza la norma l1 en la penalización a diferencia de ridge que usa la norma l2. 

    La norma de l1 de un un vector de coeficientes $\beta$ está dada por $\lVert\beta\rVert_1 = \sum \vert \beta_j \vert$




* Como en la regresión Ridge, **Lasso "achica" los coeficiente** estimados.

    En el caso de Lasso, **la regularización l1 fuerza los coeficientes a valer exactamente cero**, en el caso de que $\lambda$ sea lo suficientemente grande.

    Por lo tanto, la regularización de **Lasso hace selección de variables**

* Decimos que Lasso genera **modelos dispersos**, es decir modelos con una selección de variables

* Al igual que en Ridge, la elección de un buen valor $\lambda$ es crítico en Lasso; nuevamente **cross-validation** es el método para su elección


### Ejemplo Lasso. Dataset Credit.

<img src = "https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_credit_lasso.png" align="center" />


### Comparación Lasso Ridge

---

¿Por qué Lasso, a diferencia de Ridge, resulta en coeficientes estimados exactamente igual a cero?

La estimación de coeficientes de las regresión Lasso y Ridge resuelve estos problemas, respectivamente

<p style="font-size:20px;">
$$ minimize \space \sum_{i=1}^n (y_i - \beta_0 - \sum_{j=1}^p \beta_j x_{ij})^2  \space subject \space to \space \sum_{j=1}^p \vert \beta_j \vert \le s$$
</p>

<p style="font-size:20px;">
$$ minimize \space \sum_{i=1}^n (y_i - \beta_0 - \sum_{j=1}^p \beta_j x_{ij})^2  \space subject \space to \space \sum_{j=1}^p \beta_j ^2 \le s$$
</p>



### Lasso  gráficamente

---

<img src = "https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_lasso.png" align="center" />



### Ridge  gráficamente

---

<img src = "https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_ridge.png" align="center" />


<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_lasso_ridge.png" align="center"/>


<img src="https://github.com/Digital-House-DATA/ds_blend_2021_img/blob/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_ridgelassoItayEvron.gif?raw=true" align="center"/>

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_lasso_ridge_2.png" align="center"/>

### Comparación Lasso Ridge

---

El sesgo (bias) se calcula como la distancia desde la predicción promedio y el valor verdadero, es decir el valor verdadero menos la media de predicciones 

$Bias^2  = (E[ \hat{f}(x)]-f(x))^2$

La varianza es el desvío promedio de la predicción promedio, es decir el valor medio de (predicción menos media de predicciones) 

$Variance = E[(\hat{f}(x) - E[\hat{f}(x)])^2]$

#### Sesgo y varianza en función de $\lambda$:

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_bias_variance_regularization.png" align="center" />



Estos datos se generaron haciendo que **todos los coeficientes** fueran **diferentes a cero**. 

En este caso, **los dos modelos tienden a performar prácticamente igual**. Ridge tiene una menor varianza y por eso parece mejorar respecto a Lasso


Izquierda: Sesgo cuadrado (negro), la varianza (verde) y el MSE (violeta) para Lasso en datos simulados.

Derecha: Comparación del Sesgo cuadrado, la varianza y el MSE, entre **Lasso (llena)** y **Ridge (punteada)**.


<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_lasso_ridge_todos_los_coeficientes.png" align="center" />


Estos datos se generaron haciendo que **sólo dos coeficientes** fueran **diferentes a cero**. 

De esta forma, vemos cómo **Lasso mejora claramente la performance con respecto a Ridge**, tanto en lo referido a variancia como a MSE.

Izquierda: Sesgo cuadrado (negro), la varianza (verde) y el MSE (violeta) para Lasso en datos simulados.

Derecha: Comparación del Sesgo cuadrado, la varianza y el MSE, entre **Lasso (llena)** y **Ridge (punteada)**.


<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_lasso_ridge_dos_coeficientes.png" align="center" />



### Elastic Net 

----

<p style="font-size:20px;">
$$ \lambda_1 . \lVert \beta \lVert_1 + \lambda_2 . \lVert \beta \rVert_2^2 = \lambda . (\lVert \beta \rVert_1 + \alpha \lVert \beta \rVert_2^2)$$
</p>

* ElasticNet **combina linealmente** lo mejor de ambos mundos. 

* El parámetro $\lambda$ regula la complejidad del modelo. El parámetro $\alpha$ regula la importancia relativa de Lasso vs. Ridge. 

* Es posible obtener soluciones parsimoniosas y bien condicionadas. 

* No free lunch!: ahora hay que calibrar dos hiperparámetros.



<div class="div-dhds-fondo-1"> Cross Validation
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_separador.png" align="center" />

</div>

### Selección de Hiperparametros para Ridge y Lasso

---

* Se necesita un método para poder **seleccionar el mejor valor para el hiperparámetro $\lambda$**

* **Cross-validation** es una manera simple de atacar este problema. 
    Se elige un rango de valores que puede tomar el hiperparámetro, y se computan los errores que devuelve cross-validation para cada valor en el rango.

* Se elige el hiperparámetro asociado al menor error computado.

* Finalmente, reentrenamos el modelo con el valor elegido para el hiperparámetro .


### Validación Cruzada. Repaso

---

¿Cómo funciona?

1. Hacemos el split train/validación y test

2. Dividimos el dataset de train/validación en k grupos (generalmente, 5 o 10 suele ser la medida convencional) del mismo tamaño.

3. En la iteración i-ésima, el i-esimo grupo generado funciona como un conjunto de validación; el resto de los grupos es el conjunto de training

4. Entrenamos un modelo sobre el conjunto de training definido en 3.

5. Hacemos las predicciones sobre el conjunto de validación definido en 3. y calculamos el error sobre este conjunto

6. Repetimos los pasos 3. a 5. k veces, variando así los conjuntos de entrenamiento y de validación en cada iteración.

7. Promediamos los k errores obtenidos (un valor de error en cada una de las iteraciones).




<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_cross_validation.png" align="center" />

### Ejemplo

----

Queremos predecir el valor de la variable `Balance` en el dataset Credit (https://www.kaggle.com/avikpaul4u/credit-card-balance).

Para eso vamos a usar como variables predictoras `Income` `Limit` `Rating` `Age` y regularización de Ridge.

Vamos a sumar como variable predictora a `Rating_2` que es el resultado de 2 * `Rating` para ejemplificar el efecto de la regularización sobre un predictor colineal (recordemos los supuestos de colinealidad y regresión lineal)

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html

¿Cuál es el mejor valor de $\lambda$ (o $\alpha$)?

¿Cuál es el valor de $R^2$ del modelo en test?

Escribamos la ecuación que predice el valor de `Balance`

In [3]:
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


In [4]:
data = pd.read_csv('../Data/Credit.csv')

data['Rating_2'] = 2 * data.Rating

X = data[['Income', 'Limit', 'Rating', 'Age', 'Rating_2']]

scaler = StandardScaler()

X_std = scaler.fit_transform(X)

y = data['Balance']

In [5]:

X_train, X_test, y_train, y_test = train_test_split(X_std, y, test_size = 0.3, random_state = 117)

model_ridge_cv = linear_model.RidgeCV(alphas= [0.3, 0.5, 1.0, 1.1, 1.15, 1.17, 1.18, 1.19, 1.2, 1.21, 1.22, 1.3, 1.4, 1.5, 10.0], 
                                   fit_intercept=True, normalize=False, cv=10)

model_fit_ridge_cv = model_ridge_cv.fit(X_train, y_train)

print(model_fit_ridge_cv.alpha_)

print(model_fit_ridge_cv.best_score_)


1.18
0.8620335336980528


In [6]:
best_alpha = model_fit_ridge_cv.alpha_
model_ridge = linear_model.Ridge(alpha = best_alpha, fit_intercept = True, normalize = False)
model_fit_ridge = model_ridge.fit(X_train, y_train)
print(model_fit_ridge.coef_)
print(model_fit_ridge.intercept_)
print(model_fit_ridge.score(X_train, y_train))

[-258.48007026  148.8401642   224.7354755   -12.91648578  224.7354755 ]
512.1779140289367
0.8839840155578678


In [7]:
model_fit_ridge.score(X_test, y_test)

0.8609969065160658

El mejor valor de $\lambda$ es 1.18

El valor de $R^2$ en test es 0.86

La ecuación que predice el valor de `Balance` es

$$Balance = 512.18 -258.48 . \tilde{Income} + 148.84 .\tilde{Limit} +  224.74 . \tilde{Rating}  -12.92 . \tilde{Age} + 224.74 . \tilde{Rating\_2}$$




<div class="div-dhds-fondo-1"> Conclusión
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_separador.png" align="center" />

</div>

### Conclusiones

---

* La **regularización nos ayuda a evitar el sobreajuste** limitando la complejidad del modelo

* Matemáticamente lo logra **penalizando la complejidad** dentro de la función de costo

* Modelos con regularización suelen tener **mayor poder de generalización**

* Para determinar el valor de los **hiperparámetros** usados para regularizar, usamos **validación cruzada** 


<div class="div-dhds-fondo-1"> Hands-on
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_separador.png" align="center" />

</div>

### Hands-on

----

**Repitamos el caso del ejemplo usando Lasso:**

Queremos predecir el valor de la variable `Balance` en el dataset Credit (https://www.kaggle.com/avikpaul4u/credit-card-balance).

Para eso vamos a usar como variables predictoras `Income` `Limit` `Rating` `Age` y regularización de Ridge.

Vamos a sumar como variable predictora a `Rating_2` que es el resultado de 2 * `Rating` para ejemplificar el efecto de la regularización sobre un predictor colineal (recordemos los supuestos de colinealidad y regresión lineal)

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LassoCV.html

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html

¿Cuál es el mejor valor de $\lambda$ (o $\alpha$)?

¿Cuál es el valor de $R^2$ del modelo en test?

Escribamos la ecuación que predice el valor de `Balance`

Comparemos los resultados de Lasso con los valores de $R^2$ y los coeficientes obtenidos con un modelo de regresión lineal múltiple sin regularización.
    

### Solución

---


In [8]:
data = pd.read_csv('../Data/Credit.csv')

data['Rating_2'] = 2 * data.Rating

X = data[['Income', 'Limit', 'Rating', 'Age', 'Rating_2']]

scaler = StandardScaler()

X_std = scaler.fit_transform(X)

y = data['Balance']

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X_std, y, test_size = 0.3, random_state = 117)

model_lasso_cv = linear_model.LassoCV(alphas= [0.05, 0.06, 0.07, 0.08, 0.09, 0.95, 0.97, 0.1, 0.12, 0.15, 0.2, 0.3, 1.0, 10.0], 
                                   fit_intercept=True, normalize=False, cv=10, max_iter=1e4)

model_fit_lasso_cv = model_lasso_cv.fit(X_train, y_train)

print(model_fit_lasso_cv.alpha_)

print(model_fit_lasso_cv.score(X_train, y_train))


0.1
0.8841082428543836


In [10]:
best_alpha = model_fit_lasso_cv.alpha_
model_lasso = linear_model.Lasso(alpha = best_alpha, fit_intercept = True, normalize = False)
model_fit_lasso = model_lasso.fit(X_train, y_train)
print(model_fit_lasso.coef_)
print(model_fit_lasso.intercept_)
print(model_fit_lasso.score(X_test, y_test))

[-262.78233322  106.98310457  493.91231515  -12.87524297    1.40889939]
512.1739688178859
0.860744884074875


El mejor valor de $\lambda$ es 0.1

El valor de $R^2$ en test es 0.86

La ecuación que predice el valor de `Balance` es

$$Balance = 512.17 - 262.78 . \tilde{Income} + 106.98 . \tilde{Limit}  + 493.91 . \tilde{Rating} - 12.88 . \tilde{Age}  + 1.41 . \tilde{Rating\_2}$$



Comparemos los resultados de Lasso con los valores de $R^2$ y los coeficientes obtenidos con un modelo de regresión lineal múltiple sin regularización.

In [11]:
model = linear_model.LinearRegression()

model_fit = model.fit(X_train, y_train)
print(model_fit.coef_)
print(model_fit.intercept_)
print(model_fit.score(X_test, y_test))


[-263.21033849   99.64521564  251.53270497  -13.00094277  251.53270497]
512.1631513009786
0.8606693539960668


Observemos cómo se redujo el valor del coeficiente de `Rating_2`

<div class="div-dhds-fondo-1"> Referencias y Material Adicional
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M3/CLASE_23_Regularizacion/Presentacion/img/M3_CLASE_23_separador.png" align="center" />

</div>

### Referencias y Material Adicional

---

<a href="https://www.kaggle.com/avikpaul4u/credit-card-balance" target="_blank">Credit Card Balance</a>

<a href="https://www.statlearning.com/" target="_blank">An Introduction to Statistical Learning</a>

<a href="https://towardsdatascience.com/what-is-regularization-and-how-do-i-use-it-f7008b5a68c6" target="_blank"> Regularization with Ridge, Lasso, and Elastic Net Regressions</a>

<a href="https://people.eecs.berkeley.edu/~jrs/189/" target="_blank">L1 L2 solutions for least squares</a>


<a href="https://www.youtube.com/watch?v=Xm2C_gTAl8c" target="_blank">Ridge vs Lasso Regression, Visualized!!!</a>

<a href="https://towardsdatascience.com/bias-and-variance-in-linear-models-e772546e0c30" target="_blank">Bias and variance in linear models</a>